kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/rbac/AccessControlService.java

```diff
@@ -0,0 +1,384 @@
+    }
+
+    var mapper = YAMLMapper
+        .builder()
+        .enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS)
+        .findAndAddModules()
+        .build();
+
+    try {
+      //@formatter:off
+      this.roles = mapper.readValue(rolesFilePath.toFile(), new TypeReference<>() {});
+      //@formatter:on
+
+      this.extractors = roles
+          .stream()
+          .map(role -> role.getSubjects()
+              .stream()
+              .map(provider -> {
+                switch (provider.getProvider()) {
+                  case OAUTH_COGNITO:
+                    return new CognitoAuthorityExtractor();
+                  case OAUTH_GOOGLE:
+                    return new GoogleAuthorityExtractor();
+                  case OAUTH_GITHUB:
+                    return new GithubAuthorityExtractor();
+                  case LDAP:
+                  case LDAP_AD:
+                    return new LdapAuthorityExtractor();
+                  default:
+                    return null;
+                }
+              }).collect(Collectors.toSet()))
+          .flatMap(Set::stream)
+          .collect(Collectors.toSet());
+
+    } catch (IOException e) {
+      throw new RuntimeException("Can't deserialize roles file", e);
+    }
+
+  }
+
+  public AccessContext.AccessContextBuilder builder() {
+    return AccessContext.builder();
+  }
+
+  public Mono<Void> validateAccess(AccessContext context) {
+    return getCachedUser()
+        .doOnNext(user -> {
+          boolean accessGranted =
+              isClusterAccessible(context, user)
+                  && isTopicAccessible(context, user)
+                  && isConsumerGroupAccessible(context, user)
+                  && isConnectAccessible(context, user)
+                  && isConnectorAccessible(context, user) // TODO connector selectors
+                  && isSchemaAccessible(context, user)
+                  && isKsqlAccessible(context, user);
+
+          if (!accessGranted) {
+            throw new AccessDeniedException("Access denied");
+          }
+        })
+        .then();
+  }
+
+  public void cacheUser(AuthenticatedUser user) {
+    cachedUsers.put(user.getPrincipal(), user);
+  }
+
+  public Mono<AuthenticatedUser> getCachedUser() {
+    return ReactiveSecurityContextHolder.getContext()
+        .map(SecurityContext::getAuthentication)
+        .map(Principal::getName)
+        .flatMap(this::getCachedUser);
+  }
+
+  private Mono<AuthenticatedUser> getCachedUser(String principal) {
+    AuthenticatedUser user = cachedUsers.getIfPresent(principal);
+    if (user != null) {
+      return Mono.just(user);
+    }
+
+    return Mono.just(new AuthenticatedUser(principal, Collections.emptySet()))
+        .doOnNext(u -> cachedUsers.put(principal, u));
+  }
+
+  private boolean isClusterAccessible(AccessContext context, AuthenticatedUser user) {
+    if (context.getClusterActions().isEmpty()) {
+      log.trace("actions are empty, skipping");
```

Comment:
in what situations cluster actions will can be clear? I think at least on cluster level we require some actions

Reply:
replaced with the asserts